File: backend/src/main/kotlin/com/tnmaster/application/apis/CommonApi.kt
package com.tnmaster.application.apis

import cn.dev33.satoken.annotation.SaCheckPermission
import io.github.truenine.composeserver.data.extract.service.IMinoritiesService
import io.github.truenine.composeserver.depend.servlet.remoteRequestIp
import io.github.truenine.composeserver.generator.IOrderCodeGenerator
import io.github.truenine.composeserver.generator.ISnowflakeGenerator
import io.github.truenine.composeserver.security.crypto.CryptographicOperations
import io.github.truenine.composeserver.security.crypto.domain.IKeysRepo
import jakarta.servlet.http.HttpServletRequest
import org.babyfish.jimmer.client.ApiIgnore
import org.babyfish.jimmer.client.meta.Api
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.*

/** ## 常用数据/工具接口 */
@Api
@RestController
@RequestMapping("v2/common")
class CommonApi(
  private val minoritiesService: IMinoritiesService,
  private val bizCode: IOrderCodeGenerator,
  private val snowflakeIdGenerator: ISnowflakeGenerator,
  private val keyRepo: IKeysRepo,
) {
  /** ## 用于检测服务器是否存活 */
  @Api
  @GetMapping("ping")
  fun pong() = 1

  /** ## 获取当前用户的 ip 地址 */
  @Api
  @GetMapping("ip")
  fun getCurrentSessionRemoteIpAddr(@ApiIgnore request: HttpServletRequest): String {
    val ip = request.remoteRequestIp
    return ip
  }

  /** ## AES解密 */
  @Api
  @SaCheckPermission("ADMIN")
  @GetMapping("ass_decrypt")
  fun aseDecrypt(@RequestParam ciphertext: String): String {
    return CryptographicOperations.decryptByAesKey(keyRepo.basicAesKey()!!, ciphertext)!!
  }

  /** ## AES加密 */
  @Api
  @SaCheckPermission("ADMIN")
  @GetMapping("ase_encrypt")
  fun aseEncrypt(@RequestParam plaintext: String): String {
    return CryptographicOperations.encryptByAesKey(keyRepo.basicAesKey()!!, plaintext)!!
  }

  /** ## 获取一个 java 版的 uuid */
  @Api
  @GetMapping("java_uuid")
  fun getUuid(): String {
    return UUID.randomUUID().toString()
  }

  /** ## 雪花算法id */
  @Api
  @SaCheckPermission("ADMIN")
  @GetMapping("server_snowflake_id")
  fun getServerSnowflakeId(): String {
    return snowflakeIdGenerator.nextString()
  }

  /** ## 获取一个业务单号 */
  @Api
  @SaCheckPermission("ADMIN")
  @GetMapping("business_order_code")
  fun getBusinessOrderCode(): String {
    return bizCode.nextString()
  }

  /** ## 所有少数民族 */
  @Api
  @GetMapping("china_minorities")
  fun getAllChinaMinorities(): List<String> {
    return minoritiesService.findAllMinoritiesZ()
  }

  /**
   * ## 获取服务器异常
   * > 这通常用于测试目的
   */
  @Api
  @SaCheckPermission("ROOT")
  @GetMapping("server_throwable")
  fun getServerThrowable() {
    throw RuntimeException("服务器异常")
  }
}

File: frontend/src/__generated/services/CommonApi.ts
import type {Executor} from '../';

/**
 * ## 常用数据/工具接口
 */
export class CommonApi {
  
  constructor(private executor: Executor) {}
  
  /**
   * ## AES解密
   */
  readonly aseDecrypt: (options: CommonApiOptions['aseDecrypt']) => Promise<
    string
  > = async(options) => {
    let _uri = '/v2/common/ass_decrypt';
    let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
    let _value: any = undefined;
    _value = options.ciphertext;
    _uri += _separator
    _uri += 'ciphertext='
    _uri += encodeURIComponent(_value);
    _separator = '&';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<string>;
  }
  
  /**
   * ## AES加密
   */
  readonly aseEncrypt: (options: CommonApiOptions['aseEncrypt']) => Promise<
    string
  > = async(options) => {
    let _uri = '/v2/common/ase_encrypt';
    let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
    let _value: any = undefined;
    _value = options.plaintext;
    _uri += _separator
    _uri += 'plaintext='
    _uri += encodeURIComponent(_value);
    _separator = '&';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<string>;
  }
  
  /**
   * ## 所有少数民族
   */
  readonly getAllChinaMinorities: () => Promise<
    Array<string>
  > = async() => {
    let _uri = '/v2/common/china_minorities';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<Array<string>>;
  }
  
  /**
   * ## 获取一个业务单号
   */
  readonly getBusinessOrderCode: () => Promise<
    string
  > = async() => {
    let _uri = '/v2/common/business_order_code';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<string>;
  }
  
  /**
   * ## 获取当前用户的 ip 地址
   */
  readonly getCurrentSessionRemoteIpAddr: () => Promise<
    string
  > = async() => {
    let _uri = '/v2/common/ip';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<string>;
  }
  
  /**
   * ## 雪花算法id
   */
  readonly getServerSnowflakeId: () => Promise<
    string
  > = async() => {
    let _uri = '/v2/common/server_snowflake_id';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<string>;
  }
  
  /**
   * ## 获取服务器异常
   * > 这通常用于测试目的
   */
  readonly getServerThrowable: () => Promise<
    void
  > = async() => {
    let _uri = '/v2/common/server_throwable';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<void>;
  }
  
  /**
   * ## 获取一个 java 版的 uuid
   */
  readonly getUuid: () => Promise<
    string
  > = async() => {
    let _uri = '/v2/common/java_uuid';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<string>;
  }
  
  /**
   * ## 用于检测服务器是否存活
   */
  readonly pong: () => Promise<
    number
  > = async() => {
    let _uri = '/v2/common/ping';
    return (await this.executor({uri: _uri, method: 'GET'})) as Promise<number>;
  }
}

export type CommonApiOptions = {
  'pong': {}, 
  'getCurrentSessionRemoteIpAddr': {}, 
  'aseDecrypt': {
    ciphertext: string
  }, 
  'aseEncrypt': {
    plaintext: string
  }, 
  'getUuid': {}, 
  'getServerSnowflakeId': {}, 
  'getBusinessOrderCode': {}, 
  'getAllChinaMinorities': {}, 
  'getServerThrowable': {}
}
